src/cgame/cg_drawtools.cpp

```diff
@@ -173,22 +163,139 @@ void DrawLine(float x1, float y1, float x2, float y2, const vec4_t color) {
 
   // Use a single DrawPic for horizontal or vertical lines
   if (x1 == x2) {
-    CG_DrawPic(x1, y1 < y2 ? y1 : y2, 1, abs(y1 - y2), cgs.media.whiteShader);
+    CG_DrawPic(x1, y1 < y2 ? y1 : y2, w, std::abs(y1 - y2),
+               cgs.media.whiteShader);
   } else if (y1 == y2) {
-    CG_DrawPic(x1 < x2 ? x1 : x2, y1, abs(x1 - x2), 1, cgs.media.whiteShader);
+    CG_DrawPic(x1 < x2 ? x1 : x2, y1, std::abs(x1 - x2), h,
+               cgs.media.whiteShader);
   } else {
     len = (x2 - x1) * (x2 - x1) + (y2 - y1) * (y2 - y1);
-    len = sqrt(len);
-    stepx = (x2 - x1) / len;
-    stepy = (y2 - y1) / len;
-    for (i = 0; i < len; i++) {
-      PutPixel(x1, y1);
-      x1 += stepx;
-      y1 += stepy;
+    len = std::sqrt(len);
+    stepX = (x2 - x1) / len;
+    stepY = (y2 - y1) / len;
+    while (i < len) {
+      CG_DrawPic(x1, y1, w, h, cgs.media.whiteShader);
+      x1 += stepX;
+      y1 += stepY;
+      i++;
     }
   }
 
-  trap_R_SetColor(NULL);
+  trap_R_SetColor(nullptr);
+}
+
+struct Point {
+  float x;
+  float y;
+};
+
+static Point RotatePoint(float x, float y, float angle) {
+  float theta = DEG2RAD(angle);
+  float cos_theta = std::cos(theta);
+  float sin_theta = std::sin(theta);
+
+  float newX = x * cos_theta - y * sin_theta;
+  float newY = x * sin_theta + y * cos_theta;
+  return {newX, newY};
+}
+
+static Point TranslatePoint(Point p, float angle, const vec2_t origin) {
+  // translate triangle to the origin
+  p.x -= origin[0];
+  p.y -= origin[1];
+
+  // rotate
+  p = RotatePoint(p.x, p.y, angle);
+
+  // translate triangle back to origin
+  p.x += origin[0];
+  p.y += origin[1];
+
+  return {p.x, p.y};
+}
+
+/*
+==============
+EdgeFunction
+Edge function to determine if a point (x, y) lies inside or outside the
+triangle defined by p1, p2, and p3. The function returns a positive value if
+the point is inside the triangle, a negative value if it's outside, and zero
+if the point lies on the triangle's edge.
+==============
+*/
+static float EdgeFunction(const Point &a, const Point &b, float x, float y) {
```

Comment:
This takes a, b and x, y
Which ones are the p1-3 mentioned in the comment?
Should it be something is "isWithinTriangle"?

Reply:
Afaik "edge function" is the correct mathematical term for this calculation, hence the name, but yeah the comment is a bit weird since it mentions the points, which actually come from `FillTriangle`